modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -1242,44 +1242,44 @@ def compute_part_size(indexer, remote_part, part_idx, axis):
 
     @classmethod
     @wait_computations_if_benchmark_mode
-    def binary_operation(cls, axis, left, func, right):
+    def binary_operation(cls, left, func, right):
         """
-        Apply a function that requires two PandasDataframe objects.
+        Apply a function that requires two ``PandasDataframe`` objects.
 
         Parameters
         ----------
-        axis : {0, 1}
-            The axis to apply the function over (0 - rows, 1 - columns).
         left : np.ndarray
-            The partitions of left PandasDataframe.
+            The partitions of left ``PandasDataframe``.
         func : callable
             The function to apply.
         right : np.ndarray
-            The partitions of right PandasDataframe.
+            The partitions of right ``PandasDataframe``.
 
         Returns
         -------
         np.ndarray
             A NumPy array with new partitions.
         """
-        if axis:
-            left_partitions = cls.row_partitions(left)
-            right_partitions = cls.row_partitions(right)
-        else:
-            left_partitions = cls.column_partitions(left)
-            right_partitions = cls.column_partitions(right)
-        func = cls.preprocess_func(func)
-        result = np.array(
+        [part.drain_call_queue() for part in right.flatten()]
```

Comment:
This is the case when we can pass the right partitions and the respective call queues in the remote calls.

Reply:
We can but I don't see any benefit from applying of the respective call queue of right partition in the remote call here. This will be useful in case we have lazy execution (but even that is questionable, described below), but here we use `partition.apply` which runs an execution of remote calls right now.

In this row we run an early execution of the call queue in background for each `right_partition` and immediately actualize `right_partition._data`. After, during serial logic of `left_partition.apply` call for each partition we will have already finished remote calls from the call queue of the each `right_partition` with some probability.

If we will use your way we can than draining of call queue for `right_partition` will happen only after draining of call queue for `left_partition` in remote call.

Also, by providing the call queue of right partition together with `right_partition._data` in `left_partition.apply` we don't drain `right_partition.call_queue` and don't actualize the `right_partition._data`. If we will work with the right dataframe further in the code we will again run `drain_call_queue` for each partition of right dataframe. In the result we have double work of the same calls but in different places.

What benefits do you see from your suggestion? How we will resolve the issue which I described in this case?

Reply:
These are by far valid points but on the other hand there are others that can affect the performance. For instance, it we are considering Ray, the following factors but probably not limited to can affect the performance: 1. how many partitions are contained in the right array. 2. how many physical (materialized) data of the right partitions will be saved into in-process memory of the driver. 3. how many physical (materialized) data of the right partitions will be saved into the plasma.

We can experiment this case as part of a separate issue if we see good times without passing the call queues of the right partitions in to the remote calls.

Reply:
I still believe that the main issue with implementation you suggest will in duplicating of processing operations for right partition(the first one is draining of call queue in remote call from the left partition, the second one in right partition directly in other part of code).

I mean the next flow in the sentences above:

```python
df1 = pd.DataFrame(...) 
df2 = pd.DataFrame(...).T # Add tranpose operation in call queue of parts in df2

result = df1 + df2    # Perform cell-wise execution. Each remote call for left partition also runs transpose operation for right part. The parts of df2 isn't updated.
print(df2) # Drain call queue for parts of df2 (transpose operation)
```

So, I suggest to experiment with this new execution flow in a separate issue, because it, possible, requires a lot of architectural changes.